org.palladiosimulator.analyzer.slingshot.behavior.spd/src/org/palladiosimulator/analyzer/slingshot/behavior/spd/interpreter/utils/MedianQueue.java

```diff
@@ -0,0 +1,127 @@
+package org.palladiosimulator.analyzer.slingshot.behavior.spd.interpreter.utils;
+
+import java.util.AbstractQueue;
+import java.util.ArrayDeque;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+/**
+ * This class implements a heap-based approach of finding a median in a (bounded)
+ * queue. This queue has a maximum size, and only the last {@code n} values (according
+ * to the insertion order) will be regarded.
+ * 
+ * Here, the {@link #peek()} returns the current median, but {@link #poll()} returns
+ * the oldest element.
+ * 
+ * @author Julijan Katic
+ */
+public class MedianQueue extends AbstractQueue<Double> {
+	
+	/** Heap containing the lower half of the numbers with the maximum on the top. */
+	private final Queue<Double> maxHeap;
+	
+	/** Heap containing the higher half of the numbers with the minimum on the top. */
+	private final Queue<Double> minHeap;
+	
+	/** All the numbers added sofar in insertion order. */
+	private final Deque<Double> insertionOrderQueue;
+	
+	private final int window;
+	
+	public MedianQueue(final int windowSize) {
+		this.minHeap = new PriorityQueue<>();
+		this.maxHeap = new PriorityQueue<>(Comparator.reverseOrder());
+		this.insertionOrderQueue = new ArrayDeque<>(windowSize);
+		this.window = windowSize;
+	}
+
+	@Override
+	public boolean offer(Double e) {
+		final Double removedNumber = this.addNumber(e);
+		this.removeNumber(removedNumber);
+		
+		if (minHeap.size() == maxHeap.size()) {
+			return maxHeap.offer(e) && minHeap.offer(maxHeap.poll());
+		} else {
+			return minHeap.offer(e) && maxHeap.offer(maxHeap.poll());
```

Comment:
i get an : Filter was not successful: java.lang.NullPointerException
probably because `maxHeap.poll()` returns null, and `maxHeap.offer()` subsequently throws the NPE.